runtime/criusupport/criusupport.cpp

```diff
@@ -347,8 +351,143 @@ releaseSafeOrExcusiveVMAccess(J9VMThread *currentThread, J9InternalVMFunctions *
 	}
 }
 
+static char*
+endLine(char *cursor)
+{
+	while(true) {
+		switch (*cursor) {
+		case '\r':
+		case '\n':
+		case '\0':
+			if ((cursor[0] == '\r' && cursor[1] == '\n')
+			|| (cursor[0] == '\n')
+			) {
+				cursor[0] = '\0';
+			}
+			goto done;
+		default:
+			break;
+		}
+		cursor++;
+	}
+done:
+	return cursor;
+}
+
+static UDATA
+parseEnvVarsForOptions(J9VMThread *currentThread, J9JavaVMArgInfoList *vmArgumentsList, char *envFileChars, I_64 fileLength)
+{
+	UDATA returnCode = RESTORE_ARGS_RETURN_OK;
+	char *cursor = envFileChars;
+	char *optString = NULL;
+	bool keepSearching = false;
+
+	do {
+		keepSearching = false;
+
+		optString = strstr(cursor, OPENJ9_RESTORE_OPTS_VAR);
+
+		if (NULL != optString) {
+			/* Check that variable name is not a substring of another variable */
+			if (envFileChars != optString) {
+				char preceedingChar = *(optString - 1);
+				if ('\n' != preceedingChar) {
+					keepSearching = true;
+				}
+			}
+			optString += strlen(OPENJ9_RESTORE_OPTS_VAR);
+			cursor = optString;
+		}
+	} while (keepSearching);
+
+	if (NULL != optString) {
+		PORT_ACCESS_FROM_VMC(currentThread);
+		UDATA len = endLine(optString) - optString;
+		/* +1 for the \0 */
+		len++;
+		char *restoreArgsChars = (char*) j9mem_allocate_memory(len, OMRMEM_CATEGORY_VM);
+
+		if (NULL == restoreArgsChars) {
+			returnCode = RESTORE_ARGS_RETURN_OOM;
+			goto done;
+		}
+
+		memcpy(restoreArgsChars, optString, len);
+		currentThread->javaVM->checkpointState.restoreArgsChars = restoreArgsChars;
+
+		/* parseOptionsBuffer() will free buffer if no options were added, otherwise free at shutdown */
+		if (0 != parseOptionsBuffer(currentThread->javaVM->portLibrary, restoreArgsChars, vmArgumentsList, 0, FALSE)) {
+			returnCode = RESTORE_ARGS_RETURN_ENV_VAR_FILE_FAILED;
+		}
+	}
+done:
+	return returnCode;
+}
+
+static UDATA
+loadRestoreArgsFromEnvVariable(J9VMThread *currentThread, J9JavaVMArgInfoList *vmArgumentsList, char *envFile)
+{
+	PORT_ACCESS_FROM_VMC(currentThread);
+	IDATA fileDescriptor = -1;
+	I_64 fileLength = -1;
+	UDATA returnCode = RESTORE_ARGS_RETURN_OK;
+	char envFileBuf[ENV_FILE_BUFFER];
+	char *envFileChars = envFileBuf;
+	UDATA allocLength = 0;
+	IDATA bytesRead = 0;
+
+	fileLength = j9file_length(envFile);
+	/* Restrict file size to < 2G */
+	if (fileLength > J9CONST64(0x7FFFFFFF)) {
+		returnCode = RESTORE_ARGS_RETURN_ENV_VAR_FILE_FAILED;
+		goto done;
+	}
+
+	if ((fileDescriptor = j9file_open(envFile, EsOpenRead, 0)) == -1) {
+		returnCode = RESTORE_ARGS_RETURN_ENV_VAR_FILE_FAILED;
+		goto done;
+	}
+
+	allocLength = fileLength + 1;
+	if (allocLength > ENV_FILE_BUFFER) {
+		envFileChars = (char*) j9mem_allocate_memory(allocLength, OMRMEM_CATEGORY_VM);
+		if (NULL == envFileChars) {
+			returnCode = RESTORE_ARGS_RETURN_OOM;
+			goto close;
+		}
+	}
+
+	bytesRead = j9file_read(fileDescriptor, envFileChars, (IDATA)fileLength);
```

Comment:
Doesn't a file read typically need a loop to ensure that we've read to EOF?

Reply:
Addressed